core/src/main/java/org/elasticsearch/index/shard/ShardStateMetaData.java

```diff
@@ -100,6 +110,9 @@ public void toXContent(XContentBuilder builder, ShardStateMetaData shardStateMet
             builder.field(VERSION_KEY, shardStateMetaData.version);
             builder.field(PRIMARY_KEY, shardStateMetaData.primary);
             builder.field(INDEX_UUID_KEY, shardStateMetaData.indexUUID);
+            if (shardStateMetaData.allocationId != null) {
```

Comment:
I agree that in the current calling context (IndexShard.persistMetadata with newRouting.active()) allocationId is always non-null. 
The reasoning behind putting != null check in toXContent was to keep write and read methods aligned (so that we can write what we read - even if there is no specific use-case for that now).